h2/src/main/org/h2/mvstore/MVStore.java

```diff
@@ -357,6 +384,55 @@ private void scrubMetaMap() {
         }
     }
 
+    public void addRestorePoint(long databaseVersion, RestorePoint rp) {
+        checkNotClosed();
+        restorePoints.setWriteVersion(currentVersion);
+        restorePoints.put(databaseVersion, rp);
+    }
+
+    public void removeRestorePoint(long databaseVersion) {
+        checkNotClosed();
+        restorePoints.setWriteVersion(currentVersion);
+        restorePoints.remove(databaseVersion);
+    }
+
+    long[] getRestorePointVersions() {
+        return restorePoints
+                .keySet()
+                .stream()
+                .mapToLong(Long::longValue)
+                .sorted()
+                .toArray();
+    }
+
+    public long getOldestRestorePointVersion() {
+        return restorePoints
+                .values()
+                .stream()
+                .mapToLong(it -> it.getOldestDatabaseVersionToKeep().getLong())
+                .findFirst()
+                .orElse(Long.MAX_VALUE);
+    }
+
+    public RestorePoint findRestorePoint(String name) {
+        return restorePoints
+                .values()
+                .stream()
+                .filter(it -> it.getName().equals(name))
+                .findFirst()
+                .orElse(null);
+    }
+
+    public Collection<RestorePoint> getRestorePoints() {
+        return restorePoints.values();
+    }
+
+    public boolean isEffectivelyAppendOnly() {
```

Comment:
Such logic isn't suitable for any non-experimental usage. You already have a map with versions, so you should prevent garbage collection of these versions only, other old versions should be collectable.